server/src/main/java/org/elasticsearch/common/time/EpochTime.java

```diff
@@ -169,10 +192,18 @@ public TemporalAccessor resolve(
                 nanos = secondsAndMillis % 1000 * 1_000_000;
                 // `secondsAndMillis < 0` implies negative timestamp; so `nanos < 0`
                 if (nanosOfMilli != null) {
-                    // aggregate fractional part of the input; subtract b/c `nanos < 0`
-                    nanos -= nanosOfMilli;
+                    if (roundUp) {
```

Comment:
If we make nanos always positive as suggested above, I don't think it matters whether nanos was defaulted or came from the parse, we should always add it? Then the original "nanos != 0" branch below should work as intended, and I don't think we need roundUp at all?

Reply:
here's what I tried based on what you were suggesting and this causes failures in the parsing, which I need to look into some more (try to do that if I can get a bit of time between now and Monday): 

```
@@ -187,23 +187,14 @@ class EpochTime {
             long seconds;
             long nanos;
             if (isNegative != null) {
+                nanos = secondsAndMillis % 1000 * 1_000_000;
                 secondsAndMillis = -secondsAndMillis;
                 seconds = secondsAndMillis / 1_000;
-                nanos = secondsAndMillis % 1000 * 1_000_000;
                 // `secondsAndMillis < 0` implies negative timestamp; so `nanos < 0`
                 if (nanosOfMilli != null) {
-                    if (roundUp) {
-                        // these are not the nanos you think they are; these are "round up nanos" not the fractional part of the input
-                        // this is the case where we defaulted the value to 999_999 and the intention for rounding is that the value
-                        // moves closer to positive infinity
-                        nanos += nanosOfMilli;
-                    } else {
-                        // aggregate fractional part of the input; subtract b/c `nanos < 0`
-                        // this is the case where the user has supplied a nanos value and we'll want to shift toward negative infinity
                         nanos -= nanosOfMilli;
-                    }
                 }
-                if (nanos < 0) {
+                if (nanos != 0) {
                     // nanos must be positive. B/c the timestamp is represented by the
                     // (seconds, nanos) tuple, seconds moves 1s toward negative-infinity
                     // and nanos moves 1s toward positive-infinity
```

Reply:
to clarify I tried both `nanos += nanosOfMilli;` and `nanos -= nanosOfMilli;`

Reply:
Looked into this some more (and will probably spend a little more time tomorrow to better explain it), but the problem with this suggestion is that something like -123.123456 (without rounding) gets an incorrect value of nanos without the above selection logic with a negative nanos essentially it adds nanos when it should have subtracted.  Probably easier to walk through it together.

Reply:
discussed w @rjernst a good bit here and the game plan right now is to hold this PR open for a couple of days to noodle on any clean up / optimization to this code and otherwise to move forward with what we have here.